File: hero-csharp/Application/Models/Hero.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace hero_csharp.Models;

public class Hero
{
    public int Id { get; protected set;}
    public string Name { get; protected set;}

    public Hero(int id, string name)
    {
        Id = id;
        ChangeName(name);
    }

    public static Hero Create(string name)
    {
        return new Hero(0, name);
    }

    public void ChangeName(string name)
    {
        if(string.IsNullOrWhiteSpace(name))
        {
            throw new InvalidOperationException("Name can not be empty.");
        }

        if(name.Length < 3)
        {
            throw new InvalidOperationException("Name should have at least 3 characters.");
        }

        Name = name;

    }
}

File: hero-go/models/hero.go
package models

import (
	"errors"
	"strings"
)

type Hero struct {
	Id   int    `json:"id" gorm:"primary_key"`
	Name string `json:"name"`
}

func (Hero) TableName() string {
	// Set Table Name as heroes
	return "heroes"
}

func NewHero(name string) (*Hero, error) {
	var hero = Hero{Id: 0}
	var err = hero.ChangeHeroName(name)

	if err != nil {
		return nil, err
	}

	return &hero, nil
}

func (hero *Hero) ChangeHeroName(name string) error {
	var heroName = strings.TrimSpace(name)

	if len(heroName) == 0 {
		return errors.New("Name cannot be empty")
	}

	if len(heroName) < 3 {
		return errors.New("Name should have at least 3 characters")
	}

	hero.Name = heroName
	return nil
}
